doc/api/stream.md

```diff
@@ -90,60 +97,224 @@ server.listen(1337);
+* [fs write streams][]
 * [zlib streams][zlib]
 * [crypto streams][crypto]
+* [TCP sockets][]
+* [child process stdin][]
+* [`process.stdout`][], [`process.stderr`][]
 
-### Class: stream.Readable
+All [Writable][] streams implement the interface defined by the
+`stream.Writable` class.
+
+While specific instances of [Writable][] streams may differ in various ways,
+all Writable streams follow the same fundamental usage pattern as illustrated
+in the example below:
+
+```js
+const myStream = getWritableStreamSomehow();
+myStream.write('some data');
+myStream.write('some more data');
+myStream.end('done writing data');
+```
+
+#### Class: stream.Writable
 
 <!--type=class-->
 
-The Readable stream interface is the abstraction for a *source* of
-data that you are reading from. In other words, data comes *out* of a
-Readable stream.
+##### Event: 'close'
 
-A Readable stream will not start emitting data until you indicate that
-you are ready to receive it.
+The `'close'` event is emitted when the stream and any of its underlying
+resources (a file descriptor, for example) have been closed. The event indicates
+that no more events will be emitted, and no further computation will occur.
 
-Readable streams have two "modes": a **flowing mode** and a **paused
-mode**. When in flowing mode, data is read from the underlying system
-and provided to your program as fast as possible. In paused mode, you
-must explicitly call [`stream.read()`][stream-read] to get chunks of data out.
-Streams start out in paused mode.
+Not all Writable streams will emit the `'close'` event.
 
-**Note**: If no data event handlers are attached, and there are no
-[`stream.pipe()`][] destinations, and the stream is switched into flowing
-mode, then data will be lost.
+##### Event: 'drain'
 
-You can switch to flowing mode by doing any of the following:
+If a call to [`stream.write(chunk)`][stream-write] returns `false`, the
+`'drain'` event will be emitted when it is appropriate to resume writing data
+to the stream.
 
-* Adding a [`'data'`][] event handler to listen for data.
-* Calling the [`stream.resume()`][stream-resume] method to explicitly open the
-  flow.
-* Calling the [`stream.pipe()`][] method to send the data to a [Writable][].
+```js
+// Write the data to the supplied writable stream one million times.
+// Be attentive to back-pressure.
+function writeOneMillionTimes(writer, data, encoding, callback) {
+  var i = 1000000;
+  write();
+  function write() {
+    var ok = true;
+    do {
+      i--;
+      if (i === 0) {
+        // last time!
+        writer.write(data, encoding, callback);
+      } else {
+        // see if we should continue, or wait
+        // don't pass the callback, because we're not done yet.
+        ok = writer.write(data, encoding);
+      }
+    } while (i > 0 && ok);
+    if (i > 0) {
+      // had to stop early!
+      // write some more once it drains
+      writer.once('drain', write);
+    }
+  }
+}
+```
```

Comment:
It would be useful to visualize the way this code executes — how often does it run inside the main `do` loop before it pauses the first time? After `"drain"`, how many times does `write` call `writer.write()` before backpressure is applied again?

(The broader question to be answered here is "I have a bunch of data on hand — what is the optimal way to write it to a `WritableStream`?" I believe — and @nodejs/streams can check me on this assumption — that backpressure generally only really helps if you _don't_ have all of the items in memory already, and are able to request that the thing generating the data "back off" for a little bit. If you've already _got_ all of the data on-hand, then you're not really saving any work by holding off on writing downstream.)